content/blog/2016/2016-04-07-pipeline-for-runs-on-hardware.adoc

```diff
@@ -0,0 +1,258 @@
+
+### General improvements
+
+Let's enhance the script by using several features being provided by pipeline in order to get visualization of stages, report publishing and build notifications.
+
+We also want to minimize the time being spent on the node with the attached FPGA board. 
+So we will split the bitfile generation and further runs to two different nodes in this case: a general purpose linux node, and the node with the hardware attached.
+
+You can find the resulting Pipeline script below:
+
+```groovy
+// Synthesize on any node
+def imageId="";
+node("linux") {
+  stage "Prepare environment" 
+  git url:"http://github.com/oleg-nenashev/pipeline_hw_samples";
+  // Construct the bitfile image ID from commit ID
+  sh 'git rev-parse HEAD > GIT_COMMIT'
+  imageId= "myprj-${fpgaId}-" + readFile('GIT_COMMIT').take(6)
+  
+  stage "Synthesize project"
+  sh "make FPGA_TYPE=$fpgaId synthesize_for_fpga"
+  /* We archive the bitfile before running the test, so it won't be lost it if something happens with the FPGA run stage. */
+  archive "target/image_${fpgaId}.bit"
+  stash includes: "target/image_${fpgaId}.bit", name: 'bitfile'
+}
+
+/* Run on a node with 'my_fpga' label. 
+In this example it means that the Jenkins node contains the attacked FPGA of such type.*/
+node ("linux && $fpgaId") {  
+  stage "Blast bitfile"
+  git url:"http://github.com/oleg-nenashev/pipeline_hw_samples";
+  def artifact='target/image_'+fpgaId+'.bit';
+  echo "Using ${artifact}" 
+  unstash 'bitfile'
+  sh "make FPGA_TYPE=$fpgaId impact"
+  
+  /* We run automatic tests.
+  Then we report test results from the generated JUnit report. */
+  stage "Auto Tests"
+  sh "make FPGA_TYPE=$fpgaId tests"
+  sh "perl scripts/convertToJunit.pl --from=target/test-results/* --to=target/report_${fpgaId}.xml --classPrefix=\"myprj-${fpgaId}.\"";
+  step([$class:"JUnitResultArchiver", testResults:"target/report_${fpgaId}.xml"])
+  
+  stage "Finalization"
+  sh "make FPGA_TYPE=$fpgaId flush_fpga"
+  hipchatSend("${imageId} testing has been completed");
+}
+```
+
+As you may see, the pipeline script mostly consists of various calls of command-line tools via the _sh()_ command. 
+All EDA tools provide great CLIs, so we do not need special plugins in order to invoke common operations from Jenkins.
+
+WARNING: Disclaimer. 
+Makefile above is a sample stuff for demo purposes.
+Never write such makefiles.
+
+It is possible to continue expanding the pipeline in such way.
+link:https://github.com/jenkinsci/pipeline-examples[Pipeline Examples]
+contain examples for common cases: build parallelization, code sharing between pipelines, error handling, etc.
+
+## Lessons learned
+
+During the last 2 years I've tried using Pipeline for Hardware test automation several times.
+The first attempts were not very successful, but the ecosystem has been evolving rapidly. 
+There are still several missing integrations, but I feel Pipeline becomes a really powerful tool.
```

Comment:
@oleg-nenashev I think the disconnect when I read this was the statement said "missing integrations" whereas the list below is introduced as "following improvement areas"

I suggest using the same phrase both places to indicate that one refers to the other